### PR TITLE
Add Redis backend and automatic expiry to MongoDB

### LINKED ISSUE #27: mongo setup
@c00kiemon5ter 
I have a SATOSA/Pyop setup with mongo db, but I am not sure how data expiry and aging works in such setup.

Where can I find any info about this and mongo setup? 

How data expiry/aging works in such pyop mongo setup?
e.g. token expire very fast and after it I guess we can delete them after some wait.

I found there is a expire-data function in mongo
https://docs.mongodb.com/manual/tutorial/expire-data/

Can you please help me?
Any info would be highly appreciated!

### PR DESCRIPTION
**Pull request by bajnokk**
The first of the two patches adds Redis storage backend to pyop. It also makes it possible to assign the storage wrapper automatically based on the connection URL prefix, for example `mongodb://localhost` is routed to MongoWrapper and `redis://localhost/1` is routed to RedisWrapper.

Redis supports per-record automatic expiry, therefore a TTL is added to every record on write operations.

Since expiry is also possible for MongoDB, there is a second patch that implements the automatic expiry for MongoWrapper (also fixes #27). Note that this changes the record layout (`modified_ts` becomes `last_modified`), therefore if external tools were reading the MongoDB records, these must follow the changes.

Storage tests have been extended to include the Redis implementation, too. Since the automatic setup of the MongoDB database does not work for me (perhaps the `mongod` command line arguments in `conftest.py` are outdated?), I have only tested it with a manually started MongoDB instance. Redis tests use `fakeredis`, therefore they do not need a real database.

Questions & suggestions are welcome.

### CODE REVIEW

**Comment by bajnokk** [comment]
Updated pull request with the following:
* TTL is now a constructor parameter instead of a settable property. It makes no sense to change the TTL value during the lifetime of a wrapper object.
* Fixed a minor error in `MongoDB` class which prevented the database name to be set via the URI.
* Got rid of the manual `mongod` launching, use `mongomock` instead. Now all the unit tests should run in a clear virtualenv.
* Requires recent pytest to test the TTL expiry functionality.

File: setup.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
     description='OpenID Connect Provider (OP) library in Python.',
     install_requires=[
         'oic >= 0.15.0',
-        'pymongo'
+        'pymongo',
+        'redis'
```
Comment (c00kiemon5ter) [inline setup.py:16]: I am thinking that we should put `pymongo` and `redis` as `extras_require`. Then users can choose the preferred backend/db.

Comment (bajnokk) [reply setup.py:16]: Do you want it to be two `try: import pymongo except ImportError: pass` blocks in storage.py?

Comment (c00kiemon5ter) [reply setup.py:16]: I think that is the simplest change. I'm fine with that.

File: src/pyop/storage.py [hunk 2]
```diff
@@ -56,14 +154,17 @@ def __init__(self, db_uri, db_name=None,
...
 
         self._parsed_uri = pymongo.uri_parser.parse_uri(db_uri)
 
         if self._parsed_uri.get('database') is None:
+            if db_name is None:
+                raise ValueError(
+                    "Database name must be provided either in the URI or as an argument"
+                )
             self._parsed_uri['database'] = db_name
+        self._database_name = self._parsed_uri['database']
```
Comment (c00kiemon5ter) [inline src/pyop/storage.py:167]: I can see this being a bit simpler than nested conditions.

```suggestion
        db_name = self._parsed_uri.get('database') or db_name
        if db_name is None:
            raise ValueError(
                "Database name must be provided either in the URI or as an argument"
            )
        self._database_name = self._parsed_uri['database'] = db_name
```

**Review by c00kiemon5ter — APPROVED** [review]
I like this changeset. Thank you.

File: setup.py (hunk 1 above)
Comment (bajnokk) [reply setup.py:18]: Please check the third commit in the patchset. Do you think that the documentation should be updated as well?

File: src/pyop/storage.py (hunk 2 above)
Comment (bajnokk) [reply src/pyop/storage.py]: Thanks, updated

**Comment by maxxiefjv** [comment]
Hey Both, just here to let you know that this pull request feature looks great, and is something that we are really looking forward to. We currently rely on https://pypi.org/project/redis-collections/ as storage backend. However, this doesn't support the expiration on keys as implemented in this pull request. 

If there is anything we can do to help progress this PR, please let us know. 

Thanks!

**Comment by maxxiefjv** [comment]
Added a pull request in https://github.com/bajnokk/pyop/pull/2 to add some additional fixes in two files + an update to the latest master branch of this repo.

**Comment by c00kiemon5ter** [comment]
Thanks, this is now merged; see 

- 578c7690a3e59deb99e5ebb27799cd9665446537
- a07a54bda54a79cf11c4add38a5ea55c2a94a36b
- 270ea92d879028ae395a71e464a16f7bff086fb5

File: setup.py (hunk 1 above)
Comment (c00kiemon5ter) [reply setup.py:18]: I went ahead and merged this, but adding documentation would be great ;) of course it can be done in a separate PR. Thank you!

**Pull request closed without merging**